File: programmers/lv3/이중우선순위큐.py
from heapq import heappush, heappop

def solution(operations):
    
    mmax = []
    mmin = []
    
    maxDict = dict()
    minDict = dict()
    
    for s in operations:
        a, b = s.split()
        b = int(b)
        
        if a == 'I':
            heappush(mmax, -b)
            heappush(mmin, b)
        
        elif b == -1:
            while mmin and mmin[0] in minDict:
                minDict[mmin[0]] -= 1
                if minDict[mmin[0]] == 0:
                    minDict.pop(mmin[0])
                heappop(mmin)
            if mmin:
                maxDict.setdefault(-mmin[0], 0)
                maxDict[-heappop(mmin)] += 1
                
        elif b == 1:
            while mmax and mmax[0] in maxDict:
                maxDict[mmax[0]] -= 1
                if maxDict[mmax[0]] == 0:
                    maxDict.pop(mmax[0])
                heappop(mmax)
            if mmax:
                minDict.setdefault(-mmax[0], 0)
                minDict[-heappop(mmax)] += 1

                
    while mmin and mmin[0] in minDict:
        minDict[mmin[0]] -= 1
        if minDict[mmin[0]] == 0:
            minDict.pop(mmin[0])
        heappop(mmin)
    
    while mmax and mmax[0] in maxDict:
        maxDict[mmax[0]] -= 1
        if maxDict[mmax[0]] == 0:
            maxDict.pop(mmax[0])
        heappop(mmax)
    
    answer = [0, 0]
    
    if mmax:
        answer[0] = -mmax[0]
    if mmin:
        answer[1] = mmin[0]

    return answer

File: programmers/lv3/이중우선순위큐.java
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;

class Solution {

    static class Node implements Comparable<Node> {
        int id;
        int val;

        Node(int id, int val) {
            this.id = id;
            this.val = val;
        }

        @Override
        public int compareTo(Node node) {
            return this.val - node.val;
        }
    }

    public int[] solution(String[] operations) {
        Set<Integer> set = new HashSet<>();
        PriorityQueue<Node> maxQ = new PriorityQueue<>();
        PriorityQueue<Node> minQ = new PriorityQueue<>();

        for (int i = 0; i < operations.length; i++) {
            String s = operations[i];
            String[] ss = s.split(" ");
            if (ss[0].equals("I")) {
                Node node = new Node(i, Integer.parseInt(ss[1]));
                maxQ.add(new Node(i, -node.val));
                minQ.add(node);
            } else if (ss[1].equals("1") && maxQ.size() != 0) {
                while (maxQ.size() != 0 && set.contains(maxQ.peek().id)) {
                    maxQ.poll();
                }
                if (maxQ.size() != 0) {
                    set.add(maxQ.poll().id);
                }
            } else if (ss[1].equals("-1") && minQ.size() != 0) {
                while (minQ.size() != 0 && set.contains(minQ.peek().id)) {
                    minQ.poll();
                }
                if (minQ.size() != 0) {
                    set.add(minQ.poll().id);
                }
            }
        }

        int[] answer = { 0, 0 };

        while (maxQ.size() != 0 && set.contains(maxQ.peek().id)) {
            maxQ.poll();
        }
        if (maxQ.size() != 0) {
            answer[0] = -maxQ.peek().val;
        }

        while (minQ.size() != 0 && set.contains(minQ.peek().id)) {
            minQ.poll();
        }
        if (minQ.size() != 0) {
            answer[1] = minQ.peek().val;
        }

        return answer;
    }

}
